### PR TITLE
[webkitpy] Make Executive.kill_process more reliable.

### PR DESCRIPTION
**Pull request by bls1999**
#### 3f1a725ef3d3f0a0ee05b56aff41fbfdae90f34f
<pre>
[webkitpy] Make Executive.kill_process more reliable.
<a href="https://rdar.apple.com/127407344">rdar://127407344</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=273652">https://bugs.webkit.org/show_bug.cgi?id=273652</a>

Reviewed by Ryan Haddad and Jonathan Bedard.

Executive.kill_process currently tries to exit with SIGTERM once and then jumps straight to
SIGKILL. This can leave leftover processes, as SIGKILL sometimes doesn&apos;t give processes
enough time to react in the way that SIGTERM does.

* Tools/Scripts/webkitpy/common/system/executive.py:
(Executive.kill_process): Add more retries and wait 0.05 seconds after each.

Canonical link: <a href="https://commits.webkit.org/278309@main">https://commits.webkit.org/278309@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/524c1c86cd1582d4a5354dbd14a60674bbec1c83

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50121 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29412 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2414 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53379 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/811 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52424 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35628 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/412 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40910 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52220 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27093 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43137 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22001 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/49972 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24510 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/386 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8498 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46512 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/446 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54960 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25217 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/384 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48303 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26478 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43320 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47315 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11001 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27341 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26209 "Built successfully") | | |

### CODE REVIEW

File: Tools/Scripts/webkitpy/common/system/executive.py [hunk 1]
```diff
@@ -226,9 +226,14 @@ def kill_process(self, pid):
...
                     raise
-
-            # Give processes one chance to clean up quickly before exiting.
-            current_signal = signal.SIGKILL
+ 
+            # Fallback to SIGKILL before exiting.
+            if retries_left <= 0 and current_signal != signal.SIGKILL:
+                _log.error('Couldn\'t quit {} with SIGTERM, sending SIGKILL.'.format(pid))
+                current_signal = signal.SIGKILL
+                retries_left += 1
```
Comment (JonWBedard) [inline Tools/Scripts/webkitpy/common/system/executive.py]: You should explicitly set `retries_left` to `1`, this is the sort of logic that can bit us in the future if someone, for example, tries to make the retry counter go down `n` times faster in the future.

**Review by JonWBedard — COMMENTED** [review]
> Executive.kill_process currently tries to exit with SIGTERM once and then jumps straight to
> SIGKILL. This can leave leftover processes, as SIGKILL sometimes doesn't kill child processes
> in the same way SIGTERM does.

This is a bit pedantic, but the commit message here shouldn't talk about "child processes", since the primary motivation is `xcrun simctl`. Instead, I would say that we don't give processes enough time to react to `SIGTERM` before resorting to `SIGKILL`, and this can result in processes related to (note: _not_ the children of, as that means something specific that isn't true here) the process under test not being appropriately managed.

File: Tools/Scripts/webkitpy/common/system/executive.py (hunk 1 above)
Comment (bls1999) [reply Tools/Scripts/webkitpy/common/system/executive.py]: Good point, fixed in the recent changes. Thank you!

**Comment by bls1999** [comment]
> > Executive.kill_process currently tries to exit with SIGTERM once and then jumps straight to
> > SIGKILL. This can leave leftover processes, as SIGKILL sometimes doesn't kill child processes
> > in the same way SIGTERM does.
> 
> This is a bit pedantic, but the commit message here shouldn't talk about "child processes", since the primary motivation is `xcrun simctl`. Instead, I would say that we don't give processes enough time to react to `SIGTERM` before resorting to `SIGKILL`, and this can result in processes related to (note: _not_ the children of, as that means something specific that isn't true here) the process under test not being appropriately managed.

Just changed that commit message, thanks for pointing that out!

**Review by ryanhaddad — APPROVED** [review]

**Review by JonWBedard — APPROVED** [review]

**Pull request merged**